### PR TITLE
upload github user cache

### PR DESCRIPTION
**Pull request by superyyrrzz**
Add a config `UpdateRemoteUserCache` to control whether push cache back to remote. Upload at the same time while writing cache to disk.

#2812

### CODE REVIEW

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 1]
```diff
@@ -135,11 +137,20 @@ public Task SaveChanges()
...
+                });
+                var tasks = new List<Task>
+                {
+                    Task.Run(() => ProcessUtility.WriteFile(_cachePath, file)),
                 };
-                return ProcessUtility.WriteFile(_cachePath, JsonUtility.Serialize(file));
+                if (config.GitHub.UpdateRemoteUserCache && _isFromUrl)
+                {
+                    // TOOD: aware of ETag to handle conflicts
+                    tasks.Add(Task.Run(() => HttpClientUtility.PutAsync(config.GitHub.UserCache, new StringContent(file), config)));
```
Comment (herohua) [inline src/docfx/lib/github/GitHubUserCache.cs]: TODO: error handling?

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 2]
```diff
@@ -135,11 +137,20 @@ public Task SaveChanges()
 
             lock (_lock)
             {
-                var file = new GitHubUserCacheFile
+                var file = JsonUtility.Serialize(new GitHubUserCacheFile
                 {
                     Users = Users.ToArray(),
+                });
+                var tasks = new List<Task>
```
Comment (live1206) [inline src/docfx/lib/github/GitHubUserCache.cs]:
```suggestion
               var tasks = new List<Task>
                {
                    ProcessUtility.WriteFile(_cachePath, file),
                };
```

File: src/docfx/lib/github/GitHubUserCache.cs (hunk 1 above)
Comment (live1206) [inline src/docfx/lib/github/GitHubUserCache.cs]:
```suggestion
                    tasks.Add(HttpClientUtility.PutAsync(config.GitHub.UserCache, new StringContent(file), config));
```

Comment (live1206) [reply src/docfx/lib/github/GitHubUserCache.cs]: There is a `await` outside, think here we can just build the task array.

File: src/docfx/build/Build.cs [hunk 3]
```diff
@@ -39,7 +39,7 @@ public static async Task Run(string docsetPath, CommandLineOptions options, Repo
             context.WriteJson(manifest, "build.manifest");
 
             // TODO: write back to global cache
```
Comment (OsmondJiang) [inline src/docfx/build/Build.cs]: >// TODO: write back to global cache [](start = 12, length = 35)

remove?

File: src/docfx/restore/RestoreMap.cs [hunk 4]
```diff
@@ -55,15 +55,15 @@ string FindLastModifiedGitRepository()
             }
         }
 
-        public string GetFileRestorePath(string url)
+        public (string path, bool isFromUrl) GetFileRestorePath(string url)
```
Comment (OsmondJiang) [inline src/docfx/restore/RestoreMap.cs]: >GetFileRestorePath [](start = 45, length = 18)

how about `bool TryGetFileRestorePathFromUrl(string url)`?

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 5]
```diff
@@ -37,6 +38,7 @@ internal class GitHubUserCache
                    = new ConcurrentDictionary<string, Lazy<Task<(Error, string)>>>(StringComparer.OrdinalIgnoreCase);
 
         private readonly string _cachePath;
+        private readonly bool _isFromUrl;
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubUserCache.cs]: >_isFromUrl [](start = 30, length = 10)

_needUpdate? isFromUrl is missing context here

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 6]
```diff
@@ -135,11 +137,20 @@ public Task SaveChanges()
 
             lock (_lock)
             {
-                var file = new GitHubUserCacheFile
+                var file = JsonUtility.Serialize(new GitHubUserCacheFile
                 {
                     Users = Users.ToArray(),
+                });
+                var tasks = new List<Task>
+                {
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubUserCache.cs]: >                 [](start = 0, length = 16)

suggest to run these two steps sequentially, they have dependencies?

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 7]
```diff
@@ -135,11 +137,20 @@ public Task SaveChanges()
...
+                    Task.Run(() => ProcessUtility.WriteFile(_cachePath, file)),
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubUserCache.cs]: >Task.Run(() [](start = 20, length = 11)

don't use task.Run [https://stackoverflow.com/questions/11706788/should-methods-returning-taskt-always-start-the-returned-task/11707546#11707546](https://stackoverflow.com/questions/11706788/should-methods-returning-taskt-always-start-the-returned-task/11707546#11707546)

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 8]
```diff
@@ -126,20 +127,27 @@ public async Task<(Error, GitHubUser)> GetByCommit(string authorEmail, string re
...
                     Users = Users.ToArray(),
-                };
-                return ProcessUtility.WriteFile(_cachePath, JsonUtility.Serialize(file));
+                });
+            }
+            await ProcessUtility.WriteFile(_cachePath, file);
+            if (config.GitHub.UpdateRemoteUserCache && HrefUtility.IsHttpHref(config.GitHub.UserCache))
+            {
+                // TOOD: aware of ETag to handle conflicts
+                // TODO: error handling
```
Comment (yufeih) [inline src/docfx/lib/github/GitHubUserCache.cs]: can these TODOs be handled in this PR? otherwise it is very likely to have unhandled exception.

Comment (superyyrrzz) [reply src/docfx/lib/github/GitHubUserCache.cs]: OK. added error handling

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 9]
```diff
@@ -126,20 +127,34 @@ public async Task<(Error, GitHubUser)> GetByCommit(string authorEmail, string re
...
+            if (config.GitHub.UpdateRemoteUserCache && HrefUtility.IsHttpHref(url))
+            {
+                try
+                {
+                    // TOOD: aware of ETag to handle conflicts
+                    await HttpClientUtility.PutAsync(config.GitHub.UserCache, new StringContent(file), config);
+                }
+                catch (HttpRequestException ex)
+                {
+                    throw Errors.UploadFailed(url, ex.Message).ToException();
```
Comment (live1206) [inline src/docfx/lib/github/GitHubUserCache.cs]: Do we want to crush the entire build here if fail to update remote cache?

Comment (superyyrrzz) [reply src/docfx/lib/github/GitHubUserCache.cs]: Good question... `download-failed` is an error, but seems `upload-failed` doesn't as severe as it as it does not affect the main output.

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 10]
```diff
@@ -126,20 +127,37 @@ public async Task<(Error, GitHubUser)> GetByCommit(string authorEmail, string re
...
+            var url = config.GitHub.UserCache;
+            if (!config.GitHub.UpdateRemoteUserCache || !HrefUtility.IsHttpHref(url))
+            {
+                return null;
+            }
+
+            try
+            {
+                // TOOD: aware of ETag to handle conflicts
+                await HttpClientUtility.PutAsync(config.GitHub.UserCache, new StringContent(file), config);
```
Comment (live1206) [inline src/docfx/lib/github/GitHubUserCache.cs]: Shall we retry within `PutAsync` as we did in `GetAsync`? Or these two could be consolidated.

Comment (superyyrrzz) [reply src/docfx/lib/github/GitHubUserCache.cs]: Want to on hold this change as the retry may include checking etag and merge cache.

**Pull request merged**